---
title: "SmartPhos command line"
author: "Shubham Agrawal"
date: "2026-01-29"
output:
  BiocStyle::html_document:
    toc_float: true
    code_folding: "hide"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, dev = c("png","pdf"), fig.path = "./figureCommandLineEGF/")
library(MultiAssayExperiment)
devtools::load_all("/Users/shubhamagrawal/Documents/work/SmartPhos") # Option2: load SmartPhos package
library(tidyverse)
library(ggpubr)
library(ggrepel)
library(DT)
library(grid)
source("helper.R")
```

# Creating an R object from Spectronaut output files

[Skip for now]

```{r, eval=FALSE}
filetable <-  read.table(file="/Users/shubhamagrawal/Documents/work/compressed/DIA/fileTable.txt", header = TRUE)

mae <- readExperimentDIA(fileTable = filetable, 
                         annotation_col = c("treatment", "timepoint", "replicate",
                                            "sampleType"))
```

# Use the created object

```{r}
mae <- readRDS("data/maeObjNEW.Rds")
mae
```
# Preprocessing the assay, basic visualization, PCA

```{r}
se <- mae[["Phosphoproteome"]]
colData(se) <- colData(mae[, colnames(se)])
se
```

```{r}
plotIntensity(se[, se$sampleType == "Phospho"], color = "replicate") + theme_classic() +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 7),
    plot.title = element_text(hjust = 0.5, face = "bold")
  ) 
```


```{r}
newSE <- preprocessPhos(seData = se, transform = "log2", 
                        normalize = TRUE, impute = "QRILC")
newSE
```
```{r}
plotIntensity(newSE, color = "replicate") + theme_classic() +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 7),
    plot.title = element_text(hjust = 0.5, face = "bold")
  ) 
```
```{r}
plotMissing(newSE) + theme_classic() +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 7),
    plot.title = element_text(hjust = 0.5, face = "bold")
  ) 
```

# Perform PCA

```{r}
pca <- stats::prcomp(t(assays(newSE)[["imputed"]]), center = TRUE, scale. = TRUE)
```

```{r, eval=FALSE, include=FALSE}
p <- plotPCA(pca = pca, se = newSE, color = "treatment") +
  #geom_point(aes(colour = replicate), size = 4) +
  geom_point(aes(fill=treatment), 
       colour="black",pch=21, size=4) + coord_equal()
p
p <- plotPCA(pca = pca, se = newSE, color = "replicate") +
  #geom_point(aes(colour = replicate), size = 4) +
  geom_point(aes(fill=replicate), 
       colour="black",pch=21, size=4) + coord_equal()
p
```


```{r}
p <- plotPCA(pca = pca, se = newSE,
             color = "treatment",
             shape = "replicate")

p$layers[[1]]$aes_params$size   <- 3
p$layers[[1]]$aes_params$stroke <- 1.2

p + coord_equal() + theme(aspect.ratio = 1, legend.position = "right",
                          axis.text = element_text(color = "#000000")) +
  guides(
    shape = guide_legend(order = 1),
    color = guide_legend(order = 2)
    )
```

```{r}
p <- plotPCA(pca = pca, se = newSE,
             color = "replicate",
             shape = "treatment")

p$layers[[1]]$aes_params$size   <- 3
p$layers[[1]]$aes_params$stroke <- 1.2

p + coord_equal() + theme(aspect.ratio = 1, legend.position = "right",
                          axis.text = element_text(color = "#000000")) +
  guides(
    shape = guide_legend(order = 1),
    color = guide_legend(order = 2)
    )
```


# rep1 samples need to be removed or perform batch correction

```{r}
newSE <- preprocessPhos(seData = se, transform = "log2", 
                        normalize = TRUE, impute = "QRILC", 
                        removeOutlier = "rep1")
newSE
```

```{r}
pca <- stats::prcomp(t(assays(newSE)[["imputed"]]), 
                     center = TRUE, scale. = TRUE)
```

```{r, eval=FALSE, include=FALSE}
p <- plotPCA(pca = pca, se = newSE, color = "treatment") + 
  geom_point(aes(fill=treatment), 
       colour="black",pch=21, size=4) + coord_equal()
p
p <- plotPCA(pca = pca, se = newSE, color = "replicate") + 
  geom_point(aes(fill=replicate), 
       colour="black",pch=21, size=4) + coord_equal()
p
```

```{r}
p <- plotPCA(pca = pca, se = newSE,
             color = "treatment",
             shape = "replicate")

p$layers[[1]]$aes_params$size   <- 3
p$layers[[1]]$aes_params$stroke <- 1.2

p + coord_equal() + theme(aspect.ratio = 1, legend.position = "right",
                          axis.text = element_text(color = "#000000")) +
  guides(
    shape = guide_legend(order = 1),
    color = guide_legend(order = 2)
    )
```
```{r}
p <- plotPCA(pca = pca, se = newSE,
             color = "replicate",
             shape = "treatment")

p$layers[[1]]$aes_params$size   <- 3
p$layers[[1]]$aes_params$stroke <- 1.2

p + coord_equal() + theme(aspect.ratio = 1, legend.position = "right", 
                          axis.text = element_text(color = "#000000")) +
  guides(
    shape = guide_legend(order = 1),
    color = guide_legend(order = 2)
    )
```

```{r}
p <- plotHeatmap(type = "Top variant", newSE, top = 30, annotationCol = c("replicate", "treatment")) 
```
```{r,fig.width=10, fig.height=8}
g <- p$gtable

# Find row names grob
row_names <- which(g$layout$name == "row_names")
g$grobs[[row_names]]$gp <- gpar(fontsize = 8, fontface = "bold")
# Same for columns
col_names <- which(g$layout$name == "col_names")
g$grobs[[col_names]]$gp <- gpar(fontsize = 10)

grid.newpage()
grid.draw(g)
```


# Differential expression

```{r}
dea <- performDifferentialExp(se = newSE, 
                              assay = "imputed", 
                              method = "limma", 
                              condition = "treatment", 
                              reference = "1stCtrl", 
                              target = "EGF")
```

```{r}
p <- ggplot(dea$resDE, aes(x = pvalue)) +
  geom_histogram(fill = "grey", col = "blue", alpha=0.7) +
  ggtitle("P value histogram") + theme_classic() +
  theme(plot.title = element_text(hjust = 0.5)) 
  
p
```

```{r, eval=FALSE, include=FALSE}
plotHeatmap(type = "Differentially expressed", se = dea$seSub, data = dea$resDE, top = 50)
```
```{r, eval=FALSE, include=FALSE}
sePhoto <- newSE[1:600, ]
deaPhoto <- performDifferentialExp(se = sePhoto, 
                              assay = "imputed", 
                              method = "limma", 
                              condition = "treatment", 
                              reference = "1stCtrl", 
                              target = "EGF")

plotVolcanoPhoto <- function(resDE, pFilter = 0.05, fcFilter = 0.5, 
                           title = "Volcano plot") {
  # Convert the input table to a data frame and ensure the 'ID' column is of
  # type character
  dataVolcano <- data.frame(resDE)
  dataVolcano$ID <- as.character(dataVolcano$ID)
  # Categorize each gene based on the provided p-value and log2 fold-change
  # thresholds
  dataVolcano <- mutate(dataVolcano, expression = case_when(
    dataVolcano$log2FC >= as.numeric(fcFilter) &
      dataVolcano$pvalue <= as.numeric(pFilter) ~ "Up",
    dataVolcano$log2FC <= -as.numeric(fcFilter) &
      dataVolcano$pvalue <= as.numeric(pFilter) ~ "Down",
    dataVolcano$pvalue > as.numeric(pFilter) |
      (dataVolcano$log2FC < as.numeric(fcFilter) &
         dataVolcano$log2FC > -as.numeric(fcFilter)) ~ "Not Sig"
  ))
  
  # Create the volcano plot
  v <- ggplot(dataVolcano, aes(x = log2FC, y = -log10(pvalue))) +
    # Plot the points and color them based on their expression status
    geom_point(aes(color = expression, fill = expression), shape = 21, size = 3) +
    scale_color_manual(values = c("Up" = "firebrick4",
                                  "Down" = "#00002a", "Not Sig" = "#a9a9a9")) +
    scale_fill_manual(values = c("Up" = "red", "Down" = "#2776EA", "Not Sig" = "grey"))+
    # Add vertical lines for fold-change thresholds
    # geom_vline(xintercept = as.numeric(fcFilter), color = "darkgrey",
    #            linetype = "dashed") +
    # geom_vline(xintercept = -as.numeric(fcFilter), color = "darkgrey",
    #            linetype = "dashed") +
    # # Add horizontal lines for p-value thresholds
    # geom_hline(yintercept = -log10(as.numeric(pFilter)), color = "brown",
    #            linetype = "dashed") +
    # annotate(x = 2.5, y = -log10(as.numeric(pFilter)) - 0.2,
    #          label = paste("P-value = ", as.numeric(pFilter)),
    #          geom = "text", size = 4, color = "brown") +
    xlab("absolute log2(Quantity) difference") +
    ggtitle(title) +
    #theme_classic()
    theme_void() +
    theme(legend.position = "none")
    #theme(plot.title = element_text(hjust = 0.5))
  
  return(v)
}

plotVolcanoPhoto(deaPhoto$resDE, pFilter = 0.01, fcFilter = 1)
```



```{r}
pFilter = 0.01
plotVolcano(dea$resDE, pFilter = 0.01, fcFilter = 1) + theme_classic() +
  geom_hline(yintercept = -log10(as.numeric(pFilter)), color = "brown",
               linetype = "dashed") +
    annotate(x = 3.0, y = -log10(as.numeric(pFilter)) - 0.20,
             label = paste0("P-value = ", as.numeric(pFilter)),
             geom = "text", size = 3.5, color = "brown") 
```
```{r}
plotVolcanoDEA(dea$resDE, fcFilter = 1, pFilter = 0.01, usePadj = FALSE)
```


```{r}
dea$resDE
```



```{r}
intensityBoxPlot(se = dea$seSub, id = 's4971', symbol = "EGFR_S991")
```

# Time series clustering

```{r}
set.seed(12345)
library(matrixStats)

newSEts <- newSE[ , newSE$treatment == "EGF"]
assayMat <- SummarizedExperiment::assay(newSEts)

exprMat <- lapply(unique(newSEts$timepoint), function(tp) {
  rowMedians(assayMat[,newSEts$timepoint == tp])
  }) %>% bind_cols() %>% as.matrix()

rownames(exprMat) <- rownames(newSEts)
colnames(exprMat) <- unique(newSEts$timepoint)

sds <- apply(exprMat,1,sd)
varPer <- 80
exprMat <- exprMat[order(sds, decreasing = TRUE)[seq(1, varPer/100*nrow(exprMat))], ]
# only center when it's for expression
exprMat <- mscale(exprMat)
# remove NA values
exprMat <- exprMat[complete.cases(exprMat), ]


tsc <- clusterTS(x = exprMat, k = 5, pCut = 0.6)
```

```{r}
tsc$plot + theme(
  axis.text.x = element_text(size = 10), axis.text = element_text(color = "#000000"), ) 
```
```{r, include=FALSE, eval=FALSE}
tsc$plot + theme_void() + theme(legend.position = "none", strip.text = element_text(size=0)) + scale_color_gradient(low = "black", high = "green") + geom_line( aes(col = prob), alpha=1)
```


```{r, echo=FALSE, include=FALSE}
timerange <- unique(newSEts$timepoint)
p <- plotTimeSeries(newSEts, type = "expression", geneID = "s40", symbol = "RBM47_T519", condition = "treatment", treatment = "EGF", timerange = timerange)
p
lapply(p$layers, function(l) class(l$geom)[1])
p$layers[[1]]$aes_params$size <- 4
p$layers[[1]]$mapping$colour <- NULL
p$layers[[1]]$aes_params$colour <- "#c1191f"
p$layers[[2]]$aes_params$linewidth <- 1
p$layers[[2]]$mapping$colour <- NULL
p$layers[[2]]$aes_params$colour <- "#000000"
p + theme(axis.text = element_text(color = "#000000")) + scale_x_continuous(breaks = c(0,20,40,60,80,100))

```



# Enrichment analysis

```{r}
genesetPath <- system.file("shiny-app/geneset", package = "SmartPhos")
inGMT1 <- piano::loadGSC(paste0(genesetPath, "/Cancer_Hallmark.gmt"), 
                         type="gmt")
resTab <- enrichDifferential(dea = dea$resDE, type = "Pathway enrichment", 
                             gsaMethod = "PAGE", geneSet = inGMT1, 
                             statType = "stat", nPerm = 200, sigLevel = 0.05, 
                             ifFDR = FALSE)
resTab
```
```{r, include=FALSE, eval=FALSE}
clustEnr <- clusterEnrich(clusterTab = tsc$cluster, se = newSE, 
                          inputSet = inGMT1, filterP = 0.05, ifFDR = FALSE) 
clustEnr$plot + theme_light() + theme(
    axis.text.y = element_text(size = 10)
  ) 
```

```{r}
inGMT2 <- piano::loadGSC(paste0(genesetPath, "/KEGG_pathways.gmt"), 
                         type="gmt")
clustEnr <- clusterEnrich(clusterTab = tsc$cluster, se = newSE, 
                          inputSet = inGMT2, filterP = 0.01, ifFDR = FALSE)
clustEnr$plot + theme_light() + theme(axis.text = element_text(color = "#000000"))
```
```{r}
clustEnr$plot + theme_classic()
```


# Kinase activity inference

First on the differential expression analysis results

```{r}
netw <- getDecouplerNetwork("Homo sapiens")
scoreTab <- calcKinaseScore(dea$resDE, netw, statType = "stat", nPerm = 100)
plotKinaseDE(scoreTab, nTop = 10, pCut = 0.05)
```


```{r, eval=FALSE}
clusterData <- tsc$cluster[tsc$cluster$cluster == "cluster1",]
allClusterFeature <- clusterData %>%
  distinct(feature, .keep_all = TRUE) %>% .$feature
allClusterSite <- data.frame(rowData(newSEts)[allClusterFeature, "site"])
allClusterSite$feature <- allClusterFeature
clusterData <- clusterData %>%
  left_join(allClusterSite, by = "feature") %>%
  rename(site = "rowData.newSEts..allClusterFeature...site..")
scoreTab <- calcKinaseScore(clusterData, netw, statType = "stat", nPerm = 100)
plotKinaseTimeSeries(scoreTab, pCut = 0.05, clusterName = "cluster1")
```


# Phosphosites pattern

```{r}
newSEts <- addZeroTime(newSE, condition = "treatment", treat = "EGF",
                       zeroTreat = "1stCtrl",
                       timeRange = c("20min","40min","100min"))

rd <- as.data.frame(rowData(newSEts))
timerange <- unique(newSEts$timepoint)
sites <-  c("MET_Y1234",  "MET_Y1235", "MET_1003", "Shc1_Y427", "AKT3_Y312", "RPS6KB1", "EGFR_Y1092", "EGFR_Y1172", "SOS1_S1134", "MAP2K2_S226", "MAPK3_T202", "MAPK1_T190", "EGFR_Y1197", "ERBB2_S998","GAB1_Y659","RPS6_S244")

for (i in sites) {
  # condition to check if that site is present in row data
  if (i %in% rd$site) {
    # condition to check if assay doesn't have NAs (imputed assays are not allowed)
    if (!is.na(assay(newSEts)[rownames(rd[rd$site==i,]), ][1])) {
      p <- plotTimeSeries(newSEts, type = "expression", 
                        geneID = rownames(rd[rd$site==i,]), 
                        symbol = i, condition = "treatment", 
                        treatment = "EGF", timerange = timerange) +
        theme(axis.text = element_text(color = "#000000"), 
              axis.text.x = element_text(angle = 0, vjust = 0, hjust = 0.5, size=12.5)) + 
        scale_x_continuous(breaks = c(0,20,40,60,80,100))
      p$layers[[1]]$aes_params$size <- 4
      p$layers[[1]]$mapping$colour <- NULL
      p$layers[[1]]$aes_params$colour <- "#c1191f"
      p$layers[[2]]$aes_params$linewidth <- 1
      p$layers[[2]]$mapping$colour <- NULL
      p$layers[[2]]$aes_params$colour <- "#000000"
      print(p)
    }
  }
}
```

